examples/with-cookie-auth/lib/auth.js

```diff
@@ -0,0 +1,100 @@
+import Router from 'next/router'
+import axios from 'axios'
+
+axios.defaults.withCredentials = true
+
+const WINDOW_USER_SCRIPT_VARIABLE = `__USER__`
+
+const decode = ({ token }) => {
+  if (!token) {
+    return {}
+  }
+  const { email, type, } = token || {}
+  return { user: { email, type } }
+}
+
+export const getUserScript = (user) => {
+  const json = JSON.stringify(user)
+  return `${WINDOW_USER_SCRIPT_VARIABLE} = ${json};`
+}
+
+
+export const getServerSideToken = (req) => {
+  const { signedCookies } = req
+
+  if (!signedCookies) {
+    return {}
+  }
+  try {
+    return decode(signedCookies)
+  } catch (parseError) {
+    return {}
+  }
+}
+
+export const getClientSideToken = () => {
+  if (typeof window !== 'undefined') {
+    const user = window[WINDOW_USER_SCRIPT_VARIABLE] || {}
+    return { user }
+  }
+  return { user: {} }
+}
+
+const getRedirectPath = (userType) => {
+  switch (userType) {
+    case 'authenticated':
+      return '/profile'
+    default:
+      return '/login'
+  }
+}
+
+const redirect = (res, path) => {
+  if (res) {
+    res.redirect(302, path)
+    res.finished = true
+    return {}
+  }
+  Router.replace(path)
+  return {}
+}
+
+export const authInitialProps = (redirectIfAuth, secured) => async ({ req, res }) => {
+  const auth = req ? getServerSideToken(req) : getClientSideToken()
+  const current = req ? req.url : window.location.pathname
+  const user = auth.user
+  const isAnonymous = !user || user.type !== 'authenticated'
+  if (secured && isAnonymous && current !== '/login') {
+    return redirect(res, '/login')
+  }
+  if (!isAnonymous && redirectIfAuth) {
+    const path = getRedirectPath(user.type)
+    if (current !== path) {
+      return redirect(res, path)
+    }
+  }
+  return { auth }
+}
+
+export const getProfile = () => axios
```

Comment:
maybe you can do this `async` ?
```js
export const getProfile = async () => {
  const response = await axios.get('/api/profile')
  return response.data
}
```

Reply:
I could but the final JS code ends up bigger and not much less complicated...
```

var getProfile1 = exports.getProfile1 = function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
    var response;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return axios.get('/api/profile');

          case 2:
            response = _context.sent;
            return _context.abrupt('return', response.data);

          case 4:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined);
  }));

  return function getProfile1() {
    return _ref.apply(this, arguments);
  };
}();

var getProfile2 = exports.getProfile2 = function getProfile2() {
  return axios.get('/api/profile').then(function (response) {
    return response.data;
  });
};
```